crates/circuit/src/parameter_expression.rs

```diff
@@ -10,85 +10,1250 @@
             },
-            None => Ok(ParameterExpression {
-                expr: SymbolExpr::Value(symbol_expr::Value::Int(0)),
-            }),
+            uuid: uuid.clone(),
+            qpy_replay: None,
+            parameter_symbols: None,
+            parameter_vector: Some(Arc::new(vector.clone())),
+        })
+        .add_subclass(Parameter {})
+        .add_subclass(Self {});
+        ret
+    }
+}
+
+// ParameterExpression implementation for both Rust and Python
+impl ParameterExpression {
+    /// new for ParameterExpression
+    /// make new symbol with its name and uuid
+    pub fn new(name: String, uuid: Option<u128>) -> ParameterExpression {
+        let uuid = match uuid {
+            Some(u) => u,
+            None => Uuid::new_v4().as_u128(),
+        };
+        ParameterExpression {
+            expr: SymbolExpr::Symbol {
+                name: Box::new(name.clone()),
+                index: None,
+            },
+            uuid: uuid,
+            qpy_replay: None,
+            parameter_symbols: None,
+            parameter_vector: None,
         }
     }
 
-    /// create new expression as a symbol
-    #[allow(non_snake_case)]
-    #[staticmethod]
-    pub fn Symbol(name: String) -> Self {
-        // check if expr contains replacements for sympy
-        let name = name
-            .replace("__begin_sympy_replace__", "$\\")
-            .replace("__end_sympy_replace__", "$");
+    /// get uuid for this symbol
+    pub fn uuid(&self) -> u128 {
+        self.uuid
+    }
 
+    /// check if this is symbol
+    pub fn is_symbol(&self) -> bool {
+        if let SymbolExpr::Symbol { name: _, index: _ } = self.expr {
+            true
+        } else {
+            false
+        }
+    }
+
+    /// check if this is numeric
+    pub fn is_numeric(&self) -> bool {
+        if let SymbolExpr::Value(_) = self.expr {
+            true
+        } else {
+            false
+        }
+    }
+
+    /// check if ParameterVectorElement
+    pub fn is_vector_element(&self) -> bool {
+        if let SymbolExpr::Symbol { name: _, index } = &self.expr {
+            return match index {
+                Some(_) => true,
+                None => false,
+            };
+        }
+        false
+    }
+
+    /// return number of symbols in this expression
+    pub fn num_symbols(&self) -> usize {
+        self.expr.symbols().len()
+    }
+
+    /// check if the symbol is used in this expression
+    pub fn has_symbol(&self, symbol: String) -> bool {
+        self.expr.symbols_in_string().contains(&symbol)
+    }
+
+    /// return true if this is not complex number
+    pub fn is_real(&self) -> Option<bool> {
+        match self.expr.is_complex() {
+            Some(b) => Some(!b),
+            None => None,
+        }
+    }
+
+    // return merged set of parameter symbils in 2 parameters
+    fn merge_parameter_symbols(
+        &self,
+        other: &ParameterExpression,
+    ) -> Option<HashSet<Arc<ParameterExpression>>> {
```

Comment:
Perhaps you could do this?

```suggestion
    fn merge_parameter_symbols<'a>(
        &'a self,
        other: &'a ParameterExpression,
    ) -> Option<HashSet<&'a ParameterExpression>> {
```

This would give a view of the merged symbols to the caller and let them decide if they want to make an owned copy.